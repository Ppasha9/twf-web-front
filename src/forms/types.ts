export interface FormInput {
  labelTranslationId: string;
  name: string;
  inputType: "text" | "email" | "password" | "textarea";
  validation?: UseFormValidationProps;
}

export interface UseFormValidationProps {
  required?: boolean;
  minLength?: number;
  maxLength?: number;
}
