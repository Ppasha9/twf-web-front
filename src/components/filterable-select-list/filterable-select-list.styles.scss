@import "../../styles/abstracts/variables";

.select-constructor-item-list {
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
  &__items {
    flex-basis: 50%;
  }
  &__item {
    padding: 1rem;
    cursor: pointer;
    &:not(:last-child) {
      border-bottom: 1px dashed $color-primary-dark;
    }
    &:hover {
      background-color: $color-primary-light;
    }
    font-size: $sm-font-size;
  }
  &__search-field {
    display: flex;
    padding: 1rem 1rem 0 1rem;
    & > .form-group:nth-child(2) {
      position: relative;
      flex-grow: 1;
      margin-left: 1rem;
    }
    flex-basis: 100%;
  }
  &__search-icon {
    position: absolute;
    color: $color-primary-dark;
  }
  &__search-input {
    padding-left: 2.5rem !important;
  }
  &__item-name {
    font-weight: bold;
    font-size: $sm-font-size;
  }
  &__item-code,
  &__item-namespace {
    font-size: $xs-font-size;
  }
  &__sorters {
    display: flex;
    padding: 0 1rem 1rem 1rem;
    font-size: $sm-font-size;
    align-items: center;
    flex-basis: 100%;
  }
  &__sorter {
    display: flex;
    color: $color-primary-dark;
    align-items: center;
    cursor: pointer;
    font-weight: 700;
    &:not(:last-child) {
      margin-right: 1rem;
    }
  }
  &__filters {
    flex-basis: 48%;
    padding: 1rem;
  }
  &__filter {
    &:not(:last-child) {
      margin-bottom: 1rem;
    }
  }
}
