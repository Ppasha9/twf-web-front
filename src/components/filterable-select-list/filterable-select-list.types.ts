export interface FilterableSelectListItem {
  name: string;
  onSelect: () => void;
  [anotherProp: string]: any;
}

export interface FilterableSelectListProps {
  items: FilterableSelectListItem[];
  propsToFilter: string[];
}

export interface FilterableSelectListSortTokens {
  [filterProperty: string]: string[];
}
