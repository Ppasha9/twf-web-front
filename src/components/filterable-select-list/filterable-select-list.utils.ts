import {
  FilterableSelectListItem,
  FilterableSelectListSortTokens,
} from "./filterable-select-list.types";

export const getSortTokensFromItems = (
  items: FilterableSelectListItem[],
  propsToFilter: string[]
): FilterableSelectListSortTokens => {
  return Object.fromEntries(
    propsToFilter.map((filterProp: string) => {
      return [
        filterProp,
        items
          .reduce((acc: string[], item: FilterableSelectListItem): string[] => {
            return Array.isArray(item[filterProp])
              ? acc.concat(item[filterProp])
              : [...acc, item[filterProp]];
          }, [])
          .filter(
            (item: string, i: number, arr: string[]) => arr.indexOf(item) === i
          ),
      ];
    })
  );
};

export const getSelectedItems = (
  allItems: FilterableSelectListItem[],
  selectedSortTokens: FilterableSelectListSortTokens
): FilterableSelectListItem[] => {
  return allItems.filter((item: FilterableSelectListItem) => {
    return Object.keys(selectedSortTokens)
      .filter((token: string) => selectedSortTokens[token].length !== 0)
      .every((token: string) => {
        return Array.isArray(item[token])
          ? item[token].some((val: string) => {
              return selectedSortTokens[token].includes(val);
            })
          : selectedSortTokens[token].includes(item[token]);
      });
  });
};

export const isItemPropMatchingSearch = (
  itemPropValue: string | [],
  inputValue: string
): boolean => {
  return Array.isArray(itemPropValue)
    ? itemPropValue.join("").toLowerCase().includes(inputValue.toLowerCase())
    : itemPropValue.toLowerCase().includes(inputValue.toLowerCase());
};
