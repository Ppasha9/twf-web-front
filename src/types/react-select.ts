// react-select doesn't have descriptive type for options
export interface SelectOption {
  value: string | boolean;
  label: string;
}

export interface SelectInput {
  name: string;
  label: string;
  register: any;
  options: SelectOption[];
  defaultValue?: SelectOption;
  onBlur?: () => void;
  isMulti: boolean;
}
