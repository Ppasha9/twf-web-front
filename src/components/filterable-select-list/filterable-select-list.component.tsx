// libs and hooks
import React, { useState } from "react";
// lib components
import Select from "react-select";
// types
import {
  FilterableSelectListProps,
  FilterableSelectListItem,
  FilterableSelectListSortTokens,
} from "./filterable-select-list.types";
import { SelectOption } from "../../types/react-select";
// utils
import {
  getSelectedItems,
  getSortTokensFromItems,
  isItemPropMatchingSearch,
} from "./filterable-select-list.utils";
// icons
import Icon from "@mdi/react";
import { mdiMagnify } from "@mdi/js";
// styles
import "./filterable-select-list.styles.scss";

const FilterableSelectList = ({
  items,
  propsToFilter,
}: FilterableSelectListProps): JSX.Element => {
  const searchKeys: string[] = [
    ...Object.keys(items[0]).filter((prop: string) => {
      return prop !== "onSelect";
    }),
    "all",
  ];
  const allSortTokens: FilterableSelectListSortTokens = getSortTokensFromItems(
    items,
    propsToFilter
  );

  const [selectedItems, setSelectedItems] = useState<
    FilterableSelectListItem[]
  >(items);
  const [currentSearchKey, setCurrentSearchKey] = useState<string>("all");
  const [selectedSortTokens, setSelectedSortTokens] = useState<
    FilterableSelectListSortTokens
  >(Object.fromEntries(propsToFilter.map((prop: string) => [prop, []])));
  const [visibleSortTokens, setVisibleSortTokens] = useState<
    FilterableSelectListSortTokens
  >(allSortTokens);

  const onChangeSelect = (
    selectedOptions: SelectOption[],
    filterProp: string,
    propsToFilter: string[]
  ): void => {
    const newSelectedSortTokens = {
      ...selectedSortTokens,
      [filterProp]:
        selectedOptions === null
          ? []
          : selectedOptions.map((option: SelectOption) => option.value),
    };
    const newVisibleSortTokens: FilterableSelectListSortTokens = Object.fromEntries(
      propsToFilter.map((prop: string) => [prop, []])
    );
    const newSelectedItems: FilterableSelectListItem[] = getSelectedItems(
      items,
      newSelectedSortTokens
    );
    for (const tokenKey in allSortTokens) {
      if (allSortTokens.hasOwnProperty(tokenKey)) {
        allSortTokens[tokenKey].forEach((token: string) => {
          if (
            getSelectedItems(items, {
              ...newSelectedSortTokens,
              [tokenKey]: [...newSelectedSortTokens[tokenKey], token],
            }).length !== 0 &&
            getSelectedItems(items, {
              ...newSelectedSortTokens,
              [tokenKey]: [...newSelectedSortTokens[tokenKey], token],
            }).length !== getSelectedItems(items, newSelectedSortTokens).length
          ) {
            newVisibleSortTokens[tokenKey].push(token);
          }
        });
      }
    }
    const currentTokensOfVisibleItems = getSortTokensFromItems(
      newSelectedItems,
      propsToFilter
    );
    for (const tokenKey in currentTokensOfVisibleItems) {
      if (currentTokensOfVisibleItems.hasOwnProperty(tokenKey)) {
        currentTokensOfVisibleItems[tokenKey].forEach((token: string) => {
          if (
            newVisibleSortTokens.hasOwnProperty(tokenKey) &&
            !newVisibleSortTokens[tokenKey].includes(token)
          ) {
            newVisibleSortTokens[tokenKey].push(token);
          }
        });
      }
    }
    setSelectedItems(newSelectedItems);
    setSelectedSortTokens(newSelectedSortTokens);
    setVisibleSortTokens(newVisibleSortTokens);
  };

  const onChangeSearchInput = (searchKey: string, inputValue: string): void => {
    const relevantItems = getSelectedItems(items, selectedSortTokens);
    if (searchKey === "all") {
      setSelectedItems(
        relevantItems.filter((item: FilterableSelectListItem) => {
          return searchKeys
            .filter((searchKey: string) => searchKey !== "all")
            .some((itemKey: string) => {
              return isItemPropMatchingSearch(item[itemKey], inputValue);
            });
        })
      );
    } else {
      setSelectedItems(
        relevantItems.filter((item: FilterableSelectListItem) => {
          return isItemPropMatchingSearch(item[searchKey], inputValue);
        })
      );
    }
  };

  return (
    <div className="select-constructor-item-list">
      <div className="select-constructor-item-list__search-field">
        <div className="form-group">
          <select
            className="form-control select-constructor-item-list__select-input"
            value={currentSearchKey}
            onChange={(e: React.FormEvent<HTMLSelectElement>): void => {
              setCurrentSearchKey(e.currentTarget.value);
            }}
          >
            {searchKeys.map(
              (searchKey: string, i: number): JSX.Element => {
                return (
                  <option key={i} value={searchKey}>
                    {searchKey}
                  </option>
                );
              }
            )}
          </select>
        </div>
        <div className="form-group">
          <Icon
            path={mdiMagnify}
            size={1.5}
            className="select-constructor-item-list__search-icon"
          />
          <input
            type="text"
            className="form-control select-constructor-item-list__search-input"
            placeholder={`Введите ${currentSearchKey}`}
            onChange={(e: React.ChangeEvent<HTMLInputElement>): void => {
              onChangeSearchInput(currentSearchKey, e.target.value);
            }}
          />
        </div>
      </div>
      <div className="select-constructor-item-list__items">
        {selectedItems.length === 0 ? (
          <div style={{ padding: "1rem" }}>Элементы отсутствуют</div>
        ) : (
          selectedItems.map(
            (item: FilterableSelectListItem, i: number): JSX.Element => {
              return (
                <div
                  key={i}
                  className="select-constructor-item-list__item"
                  onClick={() => {
                    item.onSelect();
                  }}
                >
                  {Object.keys(item)
                    .filter((key: string) => key !== "onSelect")
                    .map((key: string) => {
                      const value = Array.isArray(item[key])
                        ? item[key].join(", ")
                        : item[key];
                      return (
                        <div key={key}>
                          <b>{key}: </b>
                          <span>{value}</span>
                        </div>
                      );
                    })}
                </div>
              );
            }
          )
        )}
      </div>
      <div className="select-constructor-item-list__filters">
        {propsToFilter.map(
          (filterProp: string, i: number): JSX.Element => {
            return (
              <div key={i} className="select-constructor-item-list__filter">
                <h2>{filterProp}</h2>
                <Select
                  classNamePrefix="react-select"
                  options={visibleSortTokens[filterProp].map(
                    (item: string) => ({
                      label: item,
                      value: item,
                    })
                  )}
                  isMulti={true}
                  onChange={(selectedTokens: any): void => {
                    onChangeSelect(selectedTokens, filterProp, propsToFilter);
                  }}
                />
              </div>
            );
          }
        )}
      </div>
    </div>
  );
};

export default FilterableSelectList;
