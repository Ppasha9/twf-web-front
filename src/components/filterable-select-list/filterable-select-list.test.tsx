// libs
import React from "react";
import { mount, ReactWrapper, shallow, ShallowWrapper } from "enzyme";
import { toMatchImageSnapshot } from "jest-image-snapshot";
import Select from "react-select/base";
// tested components
import FilterableSelectList from "./filterable-select-list.component";
// types
import { FilterableSelectListItem } from "./filterable-select-list.types";
// add toMatchImageSnapshot to Jest
expect.extend({ toMatchImageSnapshot });

const mockTasks: FilterableSelectListItem[] = [
  {
    name: "Запредельно сложная задача",
    namespace: "super_difficult",
    code: "very_difficult_task",
    taskSet: [
      "интересные задачки",
      "очень сложно",
      "зубодробительный набор",
      "поступление в университет",
    ],
    subjectType: [
      "тригонометрия",
      "множества",
      "дифференциалы",
      "производные",
      "полиномы",
      "пределы",
    ],
    onSelect: jest.fn().mockReturnValue(undefined),
  },
  {
    name: "Умеренно сложная задача",
    namespace: "little_bit_difficult",
    code: "not_so_easy_task",
    taskSet: ["интересные задачки", "поступление в университет", "на подумать"],
    subjectType: ["тригонометрия", "логарифмы", "теория вероятности"],
    onSelect: jest.fn().mockReturnValue(undefined),
  },
  {
    name: "Простота",
    namespace: "super_simple",
    code: "easy_breezy",
    taskSet: ["просто", "ЕГЭ"],
    subjectType: ["логарифмы", "производные"],
    onSelect: jest.fn().mockReturnValue(undefined),
  },
];

describe("filterableSelectList", () => {
  const shallowWrapper: ShallowWrapper = shallow(
    <FilterableSelectList
      items={mockTasks}
      propsToFilter={["namespace", "subjectType", "taskSet"]}
    />
  );
  const mountedWrapper: ReactWrapper = mount(
    <FilterableSelectList
      items={mockTasks}
      propsToFilter={["namespace", "subjectType", "taskSet"]}
    />
  );
  const selectKeyInput: ShallowWrapper = shallowWrapper.find(
    ".select-constructor-item-list__select-input"
  );
  const searchInput: ShallowWrapper = shallowWrapper.find(
    ".select-constructor-item-list__search-input"
  );
  const getItems = function <T>(wrapper: T): T {
    // @ts-ignore
    return wrapper.find(".select-constructor-item-list__item");
  };
  const getItemName = (item: ShallowWrapper | ReactWrapper): string => {
    return item.find("span").at(0).text();
  };

  it("matches snapshot", () => {
    expect(shallowWrapper).toMatchSnapshot();
  });

  it("search field finds expected items with 'all' search key", () => {
    selectKeyInput.simulate("change", {
      currentTarget: {
        value: "all",
      },
    });
    searchInput.simulate("change", {
      target: {
        value: "difficult",
      },
    });
    getItems(shallowWrapper).forEach((item: ShallowWrapper) => {
      expect(getItemName(item)).toMatch(
        new RegExp(`${mockTasks[0].name}|${mockTasks[1].name}`)
      );
    });
  });

  it("search field finds expected items with specific search key 1", () => {
    selectKeyInput.simulate("change", {
      currentTarget: {
        value: "all",
      },
    });
    searchInput.simulate("change", {
      target: {
        value: "производные",
      },
    });
    const items: ShallowWrapper = getItems(shallowWrapper);
    expect(items).toHaveLength(2);
    items.forEach((item: ShallowWrapper) => {
      expect(getItemName(item)).toMatch(
        new RegExp(`${mockTasks[0].name}|${mockTasks[2].name}`)
      );
    });
  });

  it("search field finds expected items with specific search key 2", () => {
    selectKeyInput.simulate("change", {
      currentTarget: {
        value: "namespace",
      },
    });
    searchInput.simulate("change", {
      target: {
        value: "this namespace doesn't exist",
      },
    });
    expect(getItems(shallowWrapper)).toHaveLength(0);
  });

  it("all provided filters are rendered on the page", () => {
    expect(mountedWrapper.find(Select)).toHaveLength(3);
  });

  it("selected tokens in filters provide expected items list 1", () => {
    // select super_difficult namespace
    mountedWrapper
      .find(".react-select__dropdown-indicator")
      .first()
      .simulate("mouseDown", {
        button: 0,
      });
    mountedWrapper.find(".react-select__option").first().simulate("click");
    const item: ReactWrapper = getItems(mountedWrapper);
    expect(item).toHaveLength(1);
    expect(getItemName(item)).toMatch(mockTasks[0].name);
    mountedWrapper.unmount().mount();
  });

  it("selected tokens in filters provide expected items list 2", () => {
    // select множества in subjectType filter
    mountedWrapper
      .find(".react-select__dropdown-indicator")
      .at(2)
      .simulate("mouseDown", {
        button: 0,
      });
    mountedWrapper.find(".react-select__option").at(2).simulate("click");
    // select производные in subjectType filter
    mountedWrapper
      .find(".react-select__dropdown-indicator")
      .at(2)
      .simulate("mouseDown", {
        button: 0,
      });
    mountedWrapper.find(".react-select__option").at(2).simulate("click");
    const items: ReactWrapper = getItems(mountedWrapper);
    expect(items).toHaveLength(2);
    items.forEach((item: ReactWrapper) => {
      expect(item.text()).toMatch(
        new RegExp(`${mockTasks[0].name}|${mockTasks[2].name}`)
      );
    });
    mountedWrapper.unmount().mount();
  });

  it("selected tokens in filters provide expected items list 3", () => {
    // select little_bit_difficult namespace
    mountedWrapper
      .find(".react-select__dropdown-indicator")
      .first()
      .simulate("mouseDown", {
        button: 0,
      });
    mountedWrapper.find(".react-select__option").at(2).simulate("click");
    // select super_difficult namespace
    mountedWrapper
      .find(".react-select__dropdown-indicator")
      .first()
      .simulate("mouseDown", {
        button: 0,
      });
    mountedWrapper.find(".react-select__option").first().simulate("click");
    // select на подумать in taskSet
    mountedWrapper
      .find(".react-select__dropdown-indicator")
      .last()
      .simulate("mouseDown", {
        button: 0,
      });
    mountedWrapper.find(".react-select__option").at(5).simulate("click");
    // remove namespace tokens
    mountedWrapper
      .find(".react-select__multi-value__remove")
      .at(1)
      .simulate("click");
    mountedWrapper
      .find(".react-select__multi-value__remove")
      .at(1)
      .simulate("click");
    const item: ReactWrapper = getItems(mountedWrapper);
    expect(item).toHaveLength(1);
    expect(getItemName(item)).toMatch(mockTasks[1].name);
    mountedWrapper.unmount().mount();
  });

  it("possible tokens to choose are relevant due to selected tokens", () => {
    // select на интересные задачки in taskSet
    mountedWrapper
      .find(".react-select__dropdown-indicator")
      .last()
      .simulate("mouseDown", {
        button: 0,
      });
    mountedWrapper.find(".react-select__option").at(1).simulate("click");
    // check namespace options
    mountedWrapper
      .find(".react-select__dropdown-indicator")
      .first()
      .simulate("mouseDown", {
        button: 0,
      });
    mountedWrapper
      .find(".react-select__option")
      .forEach((item: ReactWrapper) => {
        expect(item.text()).toMatch(
          new RegExp(`${mockTasks[0].namespace}|${mockTasks[1].namespace}`)
        );
      });
    mountedWrapper
      .find(".react-select__dropdown-indicator")
      .first()
      .simulate("mouseDown", {
        button: 0,
      });
    // check that taskSet options include all taskSets
    mountedWrapper
      .find(".react-select__dropdown-indicator")
      .last()
      .simulate("mouseDown", {
        button: 0,
      });
    const allTaskSets: string[] = mockTasks
      .reduce((acc: string[], item: FilterableSelectListItem) => {
        return acc.concat([...item.taskSet]);
      }, [])
      .filter((ts: string, i: number, arr: string[]) => {
        return arr.indexOf(ts) === i && ts !== "интересные задачки";
      });
    mountedWrapper.find(".react-select__option").forEach((ts: ReactWrapper) => {
      expect(allTaskSets).toContain(ts.text());
    });
  });
});
