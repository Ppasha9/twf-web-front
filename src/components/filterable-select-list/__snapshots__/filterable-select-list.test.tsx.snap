// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filterableSelectList matches snapshot 1`] = `
<div
  className="select-constructor-item-list"
>
  <div
    className="select-constructor-item-list__search-field"
  >
    <div
      className="form-group"
    >
      <select
        className="form-control select-constructor-item-list__select-input"
        onChange={[Function]}
        value="all"
      >
        <option
          key="0"
          value="name"
        >
          name
        </option>
        <option
          key="1"
          value="namespace"
        >
          namespace
        </option>
        <option
          key="2"
          value="code"
        >
          code
        </option>
        <option
          key="3"
          value="taskSet"
        >
          taskSet
        </option>
        <option
          key="4"
          value="subjectType"
        >
          subjectType
        </option>
        <option
          key="5"
          value="all"
        >
          all
        </option>
      </select>
    </div>
    <div
      className="form-group"
    >
      <Icon
        className="select-constructor-item-list__search-icon"
        color="currentColor"
        horizontal={false}
        path="M9.5,3A6.5,6.5 0 0,1 16,9.5C16,11.11 15.41,12.59 14.44,13.73L14.71,14H15.5L20.5,19L19,20.5L14,15.5V14.71L13.73,14.44C12.59,15.41 11.11,16 9.5,16A6.5,6.5 0 0,1 3,9.5A6.5,6.5 0 0,1 9.5,3M9.5,5C7,5 5,7 5,9.5C5,12 7,14 9.5,14C12,14 14,12 14,9.5C14,7 12,5 9.5,5Z"
        rotate={0}
        size={1.5}
        spin={false}
        vertical={false}
      />
      <input
        className="form-control select-constructor-item-list__search-input"
        onChange={[Function]}
        placeholder="Введите all"
        type="text"
      />
    </div>
  </div>
  <div
    className="select-constructor-item-list__items"
  >
    <div
      className="select-constructor-item-list__item"
      key="0"
      onClick={[Function]}
    >
      <div
        key="name"
      >
        <b>
          name
          : 
        </b>
        <span>
          Запредельно сложная задача
        </span>
      </div>
      <div
        key="namespace"
      >
        <b>
          namespace
          : 
        </b>
        <span>
          super_difficult
        </span>
      </div>
      <div
        key="code"
      >
        <b>
          code
          : 
        </b>
        <span>
          very_difficult_task
        </span>
      </div>
      <div
        key="taskSet"
      >
        <b>
          taskSet
          : 
        </b>
        <span>
          интересные задачки, очень сложно, зубодробительный набор, поступление в университет
        </span>
      </div>
      <div
        key="subjectType"
      >
        <b>
          subjectType
          : 
        </b>
        <span>
          тригонометрия, множества, дифференциалы, производные, полиномы, пределы
        </span>
      </div>
    </div>
    <div
      className="select-constructor-item-list__item"
      key="1"
      onClick={[Function]}
    >
      <div
        key="name"
      >
        <b>
          name
          : 
        </b>
        <span>
          Умеренно сложная задача
        </span>
      </div>
      <div
        key="namespace"
      >
        <b>
          namespace
          : 
        </b>
        <span>
          little_bit_difficult
        </span>
      </div>
      <div
        key="code"
      >
        <b>
          code
          : 
        </b>
        <span>
          not_so_easy_task
        </span>
      </div>
      <div
        key="taskSet"
      >
        <b>
          taskSet
          : 
        </b>
        <span>
          интересные задачки, поступление в университет, на подумать
        </span>
      </div>
      <div
        key="subjectType"
      >
        <b>
          subjectType
          : 
        </b>
        <span>
          тригонометрия, логарифмы, теория вероятности
        </span>
      </div>
    </div>
    <div
      className="select-constructor-item-list__item"
      key="2"
      onClick={[Function]}
    >
      <div
        key="name"
      >
        <b>
          name
          : 
        </b>
        <span>
          Простота
        </span>
      </div>
      <div
        key="namespace"
      >
        <b>
          namespace
          : 
        </b>
        <span>
          super_simple
        </span>
      </div>
      <div
        key="code"
      >
        <b>
          code
          : 
        </b>
        <span>
          easy_breezy
        </span>
      </div>
      <div
        key="taskSet"
      >
        <b>
          taskSet
          : 
        </b>
        <span>
          просто, ЕГЭ
        </span>
      </div>
      <div
        key="subjectType"
      >
        <b>
          subjectType
          : 
        </b>
        <span>
          логарифмы, производные
        </span>
      </div>
    </div>
  </div>
  <div
    className="select-constructor-item-list__filters"
  >
    <div
      className="select-constructor-item-list__filter"
      key="0"
    >
      <h2>
        namespace
      </h2>
      <StateManager
        classNamePrefix="react-select"
        defaultInputValue=""
        defaultMenuIsOpen={false}
        defaultValue={null}
        isMulti={true}
        onChange={[Function]}
        options={
          Array [
            Object {
              "label": "super_difficult",
              "value": "super_difficult",
            },
            Object {
              "label": "little_bit_difficult",
              "value": "little_bit_difficult",
            },
            Object {
              "label": "super_simple",
              "value": "super_simple",
            },
          ]
        }
      />
    </div>
    <div
      className="select-constructor-item-list__filter"
      key="1"
    >
      <h2>
        subjectType
      </h2>
      <StateManager
        classNamePrefix="react-select"
        defaultInputValue=""
        defaultMenuIsOpen={false}
        defaultValue={null}
        isMulti={true}
        onChange={[Function]}
        options={
          Array [
            Object {
              "label": "тригонометрия",
              "value": "тригонометрия",
            },
            Object {
              "label": "множества",
              "value": "множества",
            },
            Object {
              "label": "дифференциалы",
              "value": "дифференциалы",
            },
            Object {
              "label": "производные",
              "value": "производные",
            },
            Object {
              "label": "полиномы",
              "value": "полиномы",
            },
            Object {
              "label": "пределы",
              "value": "пределы",
            },
            Object {
              "label": "логарифмы",
              "value": "логарифмы",
            },
            Object {
              "label": "теория вероятности",
              "value": "теория вероятности",
            },
          ]
        }
      />
    </div>
    <div
      className="select-constructor-item-list__filter"
      key="2"
    >
      <h2>
        taskSet
      </h2>
      <StateManager
        classNamePrefix="react-select"
        defaultInputValue=""
        defaultMenuIsOpen={false}
        defaultValue={null}
        isMulti={true}
        onChange={[Function]}
        options={
          Array [
            Object {
              "label": "интересные задачки",
              "value": "интересные задачки",
            },
            Object {
              "label": "очень сложно",
              "value": "очень сложно",
            },
            Object {
              "label": "зубодробительный набор",
              "value": "зубодробительный набор",
            },
            Object {
              "label": "поступление в университет",
              "value": "поступление в университет",
            },
            Object {
              "label": "на подумать",
              "value": "на подумать",
            },
            Object {
              "label": "просто",
              "value": "просто",
            },
            Object {
              "label": "ЕГЭ",
              "value": "ЕГЭ",
            },
          ]
        }
      />
    </div>
  </div>
</div>
`;
